Refactor server & client implementation

maxalbert opened the pull request:
This is a substantial refactoring and enhancement of the server & client architecture and implementation. I'll leave the few remaining fixes as well as tidying, documentation, change log updates, etc. for subsequent PRs so that the existing enhancements are available in the dev branch.

Niolon commented:
Had a quick look this morning. At the moment you can add both required and optional commands in two ways
 1. Via Sets, You have laid the ground work for that in the database.
 2. Directly in the command definition. For an example see `check_structure_convergence` from the `config_qcrboxtools.yaml`. I thought this might be more feasible, if the definition of a set is not really necessary (because the commands do not share a lot of cif entries with each other). 
 
Also that might just be me but I am worried about naming conflicts between cif sets from different applications, or is that covered?

maxalbert committed:
Reduce output for 'qcb list commands'

maxalbert committed:
Only print names in 'parameters' column

maxalbert committed:
Merge remote-tracking branch 'origin/dev'

To fix merge conflicts, the 'call_pattern' entries in
services/applications/olex2_linux/config_olex2.yaml
were updated to match those in
services/applications/olex2_linux/configure_olex2.py.

maxalbert committed:
Update requirements files

maxalbert committed:
Set env var QCRBOX__RABBITMQ__HOST in base_ancestor image so that it is picked up by the new settings.

maxalbert committed:
Set docker-compose env var to specify server port (and fix health check)

maxalbert committed:
Fix import in prebuild script

maxalbert committed:
Add pytest-cases as dev dependency

maxalbert committed:
Add test that yaml config files can be parsed without errors

maxalbert committed:
Add option 'implemented_as: "interactive"' (disabled GUI for the time being)

maxalbert committed:
Add missing call_pattern and import_path values to config YAML files

maxalbert committed:
Use underscores in filenames to ensure .py file can be imported as a module

maxalbert committed:
Allow dicts in cif entry sets (needed to support alternative options via 'one_of')

maxalbert committed:
In configure_olex2.py, read application spec from YAML file instead of definining commands inline

maxalbert committed:
Add TODO comment to speed up docker build process

maxalbert committed:
Ensure the 'host' setting is honoured when running QCrBox server/client

maxalbert committed:
Listen on all interfaces when running inside docker

maxalbert committed:
Make 'ping' command available in base image (for debugging purposes)

maxalbert committed:
Add 'requests' as dev dependency (for pre-build scripts)

maxalbert committed:
Use proper health check for application containers

maxalbert committed:
Declare QCRBOX__REGISTRY__{CLIENT|SERVER}__{HOST|PORT} as ENV vars in Dockerfiles

maxalbert committed:
Add health-check endpoint to client API

maxalbert committed:
Replace localhost -> 127.0.0.1

maxalbert committed:
Add convenience CLI command to update dependency versions in requirements*.txt files

maxalbert committed:
Update requirements*.txt files

maxalbert committed:
Add default values for action string literals

maxalbert committed:
Add stub for updated version of qcrbox_wrapper

maxalbert committed:
Merge branch 'dev' into refactor-server-client

maxalbert committed:
Remove obsolete helper function

maxalbert committed:
Rename build arg and env var for consistency with new settings config: QCRBOX_RABBITMQ_PORT -> QCRBOX__RABBITMQ__PORT

maxalbert committed:
Allow setting docker project name in 'qcb up/down'

maxalbert committed:
Add setting for RabbitMQ management port (default: 15672)

maxalbert committed:
Rename env var: QCRBOX_RABBITMQ_MANAGEMENT_PORT -> QCRBOX__RABBITMQ__MANAGEMENT_PORT

maxalbert committed:
Rename env var: QCRBOX_RABBITMQ_HOST -> QCRBOX__RABBITMQ__HOST

maxalbert committed:
Add stamina as dependency

maxalbert committed:
Make connection attempts to RabbitMQ more robust (using stamina for retries)

maxalbert committed:
Add support for docker services without local build contexts (for which images can be pulled directly from DockerHub)

maxalbert committed:
Update yaml config: parameter.type -> parameter.dtype

maxalbert committed:
Add stub for QCrBoxCommand in updated wrapper

maxalbert committed:
Mark test for YAML config files as xfail (ApplicationSpec classes need to be updated to the new config format)

maxalbert committed:
Add stub for qcrbox_wrapper test

maxalbert committed:
New QCrBoxWrapper is initialised with a web client(of type httpx.AsyncClient) instead of server address and port

This allows for easy testing within the current implementation. Likely to
change again at a later stage.

maxalbert committed:
Test for QCrBoxWrapper properly checks for connection to the test server and validates registered dummy application

maxalbert committed:
Add helper method to instantiate QCrBoxWrapper from server_addr (and optional server_port)

maxalbert committed:
Switch web client to httpx.Client

Avoids complications with mixing sync and async code in QCrBoxWrapper

maxalbert committed:
Add synchronous web client to test server

maxalbert committed:
Add test for application_dict

maxalbert committed:
Tweak command defined in dummy application

maxalbert committed:
Allow passing application config yaml file when running pyqcrbox client from the command line

maxalbert committed:
Stub for /invoke_command API endpoint

maxalbert committed:
Add API endpoint to invoke a command by id

maxalbert committed:
Add svcs as dependency

maxalbert committed:
Initialise svcs.Registry when creating QCrBox server/client

maxalbert committed:
Rename test file

maxalbert committed:
Declare Python 3.11 as (currently) minimum supported Python version

maxalbert committed:
Add .tool-versions to .gitignore (used by asdf to manage local tool versions)

maxalbert committed:
Rename files: api.py -> api_endpoints.py

maxalbert committed:
Tweak server API version

maxalbert committed:
Add test for /invoke_command API endpoint

maxalbert committed:
Add trace-level logging message

maxalbert committed:
Rename API endpoint: /health-check -> /healthcheck

maxalbert committed:
Rename msg_spec class: ClientIsAvailableToExecuteCommand -> ClientIndicatesAvailabilityToExecuteCommand

maxalbert committed:
Tweak logging message

maxalbert committed:
Remove calls to ensure_dict() in tests

Not necessary any more now that this bug has been fixed:
https://github.com/airtai/faststream/issues/1437

maxalbert committed:
Check expected response data in test

maxalbert committed:
Define global svcs.Registry instance for qcrbox

maxalbert committed:
Print warning instead of aborting if a queue has more than one handler defined

maxalbert committed:
Send RabbitMQ message from /invoke_command endpoint

maxalbert committed:
Allow passing response payload as dict in convenience functions

maxalbert committed:
Return response message which includes correlation_id when invoking command

maxalbert committed:
Modify test to invoke command via RPC call and check for correlation_id in response

maxalbert committed:
Return correlation_id from API endpoint

maxalbert committed:
Add custom exception QCrBoxDBError

maxalbert committed:
Deal with duplicate correlation ids in CommandInvocationCreate.save_to_db()

maxalbert committed:
Save command invocation to the database before processing it further

maxalbert committed:
Also return cmd_id in response

maxalbert committed:
Add sql_model CalculationDB

maxalbert committed:
Create db entry for calculation when command is invoked

maxalbert committed:
Remove API endpoint to invoke command by id for the time being

maxalbert committed:
Alternative endpoint /commands/invoke for command invocation

maxalbert committed:
Add API endpoint /calculations/<id> to retrieve status info

maxalbert committed:
Add test for calculation status 'running'

maxalbert committed:
Add dev dependency: pytest-hot-reloading

maxalbert committed:
Add debugging message

maxalbert committed:
Slightly simplify API responses

maxalbert committed:
Update calculation status sending message to check for available clients

maxalbert committed:
Mark test as xfail

maxalbert committed:
Relax constraints on SQLiteDsn (validation is currently buggy anyway)

maxalbert committed:
Store calculation status updates in separate table (helpful to re-trace the sequence of events)

maxalbert committed:
Fix test

maxalbert committed:
Ensure that status events are always linked to their parent calculation

maxalbert committed:
Fix creation of initial status event

maxalbert committed:
Fix missing calculation_id in initial event; update test

maxalbert committed:
Fix expected response in test

maxalbert committed:
Add type hint

maxalbert committed:
Add stubs for the various executable command types (interactive, Python callable, CLI command)

maxalbert committed:
Basic implementation for PythonCallable

maxalbert committed:
Remove TODO comment

maxalbert committed:
Include traceback in logging message if the PythonCallable raises an exception during execution

maxalbert committed:
Parse/validate function argument types before execution

maxalbert committed:
Terminate running calculations if the client receives a shutdown signal

maxalbert committed:
Flesh out CLICommand

maxalbert committed:
Update QCrBoxCommand

maxalbert committed:
The terminate() method lives on CLICommand, not CLICmdCalculation

maxalbert committed:
Remove db models for command invocation and execution (superseded by CalculationDB)

maxalbert committed:
Fix model validator (need to return self)

maxalbert committed:
Fix test (need test_client so that the expected command is registered before the test runs)

maxalbert committed:
Typo

maxalbert committed:
Raise custom exception if arguments do not match command signature; fix CLICommand.terminate() implementation

maxalbert committed:
Validation code needs access to the database and shouldn't live on a pure pydantic model

maxalbert committed:
Catch message processing errors and return a proper error response

maxalbert committed:
Gracefully handle certain error corner cases

maxalbert committed:
Add explicit setting for RabbitMQ rpc_timeout

maxalbert committed:
Return response from client

maxalbert committed:
Convert response dict to QCrBoxGenericResponse to ensure the structure is correct

maxalbert committed:
Move message handler for command execution into separate file

maxalbert committed:
Decrease graceful_timeout to 5 seconds

maxalbert committed:
Add debugging message

maxalbert committed:
Initialise empty dict for custom notification events that can be used by the server (or client)

maxalbert committed:
Remove obsolete '/invoke_command' endpoint (superseded by '/commands/invoke')

maxalbert committed:
Use rpc_timeout from settings in RPC call

maxalbert committed:
Command invocation request is sent as regular message (not RPC)

Multiple clients may be receiving the invocation request so we can't use
a remote procedure call for this message. Instead, a notification event
is triggered when a client responds and starts executing the command.

maxalbert committed:
Import logger from pyqcrbox instead of loguru

maxalbert committed:
Rename method/helper function for clarity

maxalbert committed:
Allow retrieving calculation info by correlation_id

maxalbert committed:
Rename parameter: routing_key -> queue_name

maxalbert committed:
Provide optional arguments exchange_type, routing_key when attaching message dispatcher

maxalbert committed:
Rename method and helper function

maxalbert committed:
Add NATS container

maxalbert committed:
Start NatsBroker upon server/client startup

maxalbert committed:
Install nats extras for faststream

maxalbert committed:
Add nats broker in server/client base class

maxalbert committed:
Client sends registration message via NATS broker

maxalbert committed:
Server listens for application registration messages via NATS

maxalbert committed:
Server handles application registration message sent via NATS

maxalbert committed:
Skip setting up message dispatcher for RabbitMQ so we can test NATS messaging instead

maxalbert committed:
Rip out RabbitBroker and replace it with NATS-based messaging

maxalbert committed:
Reduce rpc_timeout to three seconds

maxalbert committed:
Command invocation via web API sends NATS request to available clients

maxalbert committed:
Server handles client responses to command invocation requests

maxalbert committed:
Discard spurious empty messages

These are likely an artefact/bug in FastStream but should not cause any
problems now that they are ignored.

maxalbert committed:
Use nats docker image based on alpine (to be able to run a shell inside the container for debugging)

maxalbert committed:
Switch to structlog for logging

maxalbert committed:
Introduce NATS-related env vars and use them to set ports in docker-compose.run.yml

maxalbert committed:
handle_application_registration() is now a method on QCrBoxServer

maxalbert committed:
Make set_log_level() work with structlog

maxalbert committed:
Add NATS versions of InvokeCommand messages

maxalbert committed:
Add NATS-based message handler for command invocation

maxalbert committed:
Client responds to command invocation request from server

maxalbert committed:
Server sends command execution request to client

maxalbert committed:
Switch to default ports for NATS in .env.dev

maxalbert committed:
Enable NATS monitoring

maxalbert committed:
Add healthcheck block for nats container (commented out for now because it requires curl which needs to be installed first)

maxalbert committed:
Set up private inbox in client/server base class

maxalbert committed:
Enable command execution via PythonCallable

maxalbert committed:
Calculation status can be retrieved from server (and client)

maxalbert committed:
Enable structlog plugin for server/client Litestar apps

maxalbert committed:
Fix /commands/invoke and /calculations endpoints

They now call the NATS-based message handlers under the hood (also
storing the relevant data in the SQLite database).

maxalbert committed:
Retrieve calculation status directly from NATS key-value store (lookup is now by calculation_id)

maxalbert committed:
Add nats-nui container (NATS management GUI)

maxalbert committed:
Debugging statement about sleep duration

maxalbert committed:
Helper functions to retrieve nats broker and key-value bucket

maxalbert committed:
Update status in NATS key-value store when calculation is finished

maxalbert committed:
Tidying

maxalbert committed:
More tidying; set client status back to 'idle' when calculation is done

maxalbert committed:
Update calculation status in the database when the value in the NATS KV changes

maxalbert committed:
Reduce fields returned by /calculations endpoint

maxalbert committed:
Set calculation status to 'submitted' as soon as the server receives the command invocation from the user

maxalbert committed:
If client is busy, do not process the response further

maxalbert committed:
Fix CalculationDB.status (ensure we're only considering status events corresponding to this calculation)

maxalbert committed:
Tidy up debugging messages

maxalbert committed:
Instantiate executable command from the command name in the incoming message

maxalbert committed:
Fix execution and status updates of both PythonCallable and CLICommand calculations

maxalbert committed:
Add dummy CLI command to sample_application_spec.yaml

maxalbert committed:
Slightly more graceful shutdown if server does not respond

maxalbert committed:
In case of failure during command execution, log the error and set the client status back to 'idle'

maxalbert committed:
Print stdout/stderr in case of failed command execution

maxalbert committed:
Add missing dependency 'structlog'

maxalbert committed:
Update dependency versions

maxalbert committed:
Add healthcheck to nats container (also added Dockerfile to install curl and bash)

maxalbert committed:
Fix env vars in docker-compose files

maxalbert committed:
Replace RabbitMQ dependencies in FastStream with NATS

maxalbert committed:
Disable logs for /healthcheck endpoint

maxalbert committed:
Reduce startup periods for NATS and qcrbox-registry containers

maxalbert committed:
Revert "Disable logs for /healthcheck endpoint"

This reverts commit 26339edae9b094f324d36b97bbaba53c51cd849e.

maxalbert committed:
Return JSON response from /healthcheck endpoints

maxalbert committed:
Merge branch 'dev' into refactor-server-client

maxalbert committed:
Store more details about calculation status and return this in the API endpoint

maxalbert committed:
Fix status details for PythonCallableCalculation and treat them uniformly with CLICommand

maxalbert committed:
Add jupyterlab as dev dependency

maxalbert committed:
Reduce healthcheck frequency to avoid cluttering container logs

maxalbert committed:
Update dependency versions

maxalbert committed:
Fix call to ValidateCallWrapper (requires an additional argument 'namespace' in pydantic >=2.8.0)

maxalbert committed:
Temporarily comment out --start-interval option

maxalbert committed:
Tweak error message

maxalbert committed:
Immediately respond with a client error if the invoked command does not exist

maxalbert committed:
Extract parameter specs from function signature

maxalbert committed:
Remove a bunch of obsolete code related to the old message dispatch mechanism

maxalbert committed:
Fix server responses

maxalbert committed:
Make argument 'name' mandatory in dummy function (to test error response for missing required arguments)

maxalbert committed:
Validate arguments supplied in command invocation against command parameters

maxalbert committed:
Simplified and more flexible version of application/command/parameter specs

maxalbert committed:
Add parameter validation

maxalbert committed:
Add custom parameter types for QCrBox.input_cif, QCrBox.input_file, QCrBox.output_cif

maxalbert committed:
Disallow extra fields in validation

maxalbert committed:
Rename spec classes

maxalbert committed:
WIP - sql_models_NEW

maxalbert committed:
Add new folder for tidied-up version of pydantic models related to application spec

maxalbert committed:
Add pydantic models related to command spec

maxalbert committed:
Add pydantic models for parameter types

maxalbert committed:
Allow command lookup by name

maxalbert committed:
Add convenience attribute `ApplicationSpec.cmds_by_name`

maxalbert committed:
Validate Python callable parameters against function signature

maxalbert committed:
Add interactive command spec

maxalbert committed:
Explicit python callable parameters in sampe_application_spec.yaml

maxalbert committed:
Add QCrBox-specific filesystem parameter types

maxalbert committed:
Add CifEntrySet fields

maxalbert committed:
Add qcrbox_yaml_spec_version

maxalbert committed:
Switch application registration models to new ApplicationSpec

maxalbert committed:
Tweaks to allow serialisation of ApplicationSpec object to json string

maxalbert committed:
Store ApplicationSpec in nats KV during registration

maxalbert committed:
Move pyqcrbox.svcs submodule into its own folder

maxalbert committed:
Add NatsPersistenceAdapter

maxalbert committed:
Add main guard in configure_olex2.py

maxalbert committed:
Update config_olex2.yaml to conform to latest spec format

maxalbert committed:
Add ParameterSpecDB sql model

maxalbert committed:
Upgrade dependency versions in requirements files

maxalbert committed:
Refactor parameter spec definition and improve validation

maxalbert committed:
Add CommandSpecDB model

maxalbert committed:
WIP - rename existing submodules sql_models and sql_models_NEW

maxalbert committed:
Deactivate importlinter for the time being

maxalbert committed:
Add stub for ApplicationSpecDB

maxalbert committed:
Rename CommandSpecDBNEW -> CommandSpecDB

maxalbert committed:
Flesh out ApplicationSpecDB

maxalbert committed:
Avoid circular imports

maxalbert committed:
Add ApplicationSpecDB.save_to_db()

maxalbert committed:
Add models for command invocation, execution and calculation; fix imports

maxalbert committed:
Add SQLitePersistenceAdapter

maxalbert committed:
Add qcrbox_yaml_spec_version to sample_application_spec.yaml

maxalbert committed:
Add methods to convert db models to API return models

maxalbert committed:
Persist application spec to SQLite database upon registration

maxalbert committed:
Rename healthcheck endpoints to /healthz

maxalbert committed:
Fix JSON dump for API return model

maxalbert committed:
Small fixes

maxalbert committed:
Update wrapper

maxalbert committed:
Add CalculationStatusEventDB

maxalbert committed:
Missing import

maxalbert committed:
Fix import in configure_olex2.py

maxalbert committed:
Bump DOCKER_TAG

maxalbert committed:
Merge branch 'dev' into refactor-server-client
